File: frontend/src/lib/queries/drafts.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import {
  deleteDraft,
  getDraftBoard,
  listDraftedApplications,
  makeDraftPick,
  openDraft,
  removeDraftPick,
  replaceDraftPick,
  resetDraft,
  setDraftTeams,
  updateDraft,
} from '@/generated/draft/draft'
import type { RequestOptions } from '@/lib/api/orval-mutator'
import type {
  DraftBoard,
  DraftPickDetail,
  DraftStatus,
  Role,
} from '@/lib/api/types'
import { queryKeys } from './keys'

// The board is driven by one operator in a room full of people watching it, so
// it polls on the same 8s cadence the preference lists use — everyone sees a
// pick land without reloading.
const DRAFT_POLL_MS = 8000

export function useDraftBoard(
  cycleId: string,
  role: Role,
  { poll = false }: { poll?: boolean } = {},
  opts?: RequestOptions
) {
  return useQuery({
    queryKey: queryKeys.drafts.board(cycleId, role),
    queryFn: () =>
      getDraftBoard(cycleId, { role }, opts) as Promise<DraftBoard>,
    enabled: !!cycleId,
    refetchInterval: poll ? DRAFT_POLL_MS : false,
    // A board that hasn't been opened yet 404s, which is a normal state here
    // (the setup screen offers to open it), not something to retry into.
    retry: false,
  })
}

// Application id -> the name of the team that took them, across every board in
// the cycle. Cheap enough to poll alongside a list people draft from.
export function useDraftedApplications(
  cycleId: string,
  { poll = false }: { poll?: boolean } = {},
  opts?: RequestOptions
) {
  return useQuery({
    queryKey: queryKeys.drafts.drafted(cycleId),
    queryFn: async () =>
      ((await listDraftedApplications(cycleId, opts)) ?? {}) as Record<
        string,
        string
      >,
    enabled: !!cycleId,
    refetchInterval: poll ? DRAFT_POLL_MS : false,
  })
}

// Which team holds a slot. Filling or emptying a slot never changes the
// mapping — a pick is stamped with the team the mapping already gave it — so
// the board's own slots stay valid across an optimistic update.
export function ownerOfSlot(board: DraftBoard, slot: number) {
  return board.slots?.find((s) => s.pick_number === slot)?.draft_team_id
}

// The lowest unfilled slot is what's on the clock, so undoing a pick mid-board
// puts that slot back rather than the end of the board. Mirrors nextOpenSlot
// and DraftBoard in the backend's store/drafts.go — a predicted board that
// disagrees would name the wrong team until the refetch corrects it. Exported
// for the test that holds it to those two.
export function withOnTheClock(board: DraftBoard): DraftBoard {
  const filled = new Set(board.picks.map((p) => p.pick_number))
  let slot = 0
  for (let n = 1; n <= board.rounds * board.teams.length; n++) {
    if (!filled.has(n)) {
      slot = n
      break
    }
  }
  if (slot === 0 || board.status !== 'active' || board.teams.length === 0) {
    return { ...board, on_the_clock: 0, on_the_clock_team_id: undefined }
  }
  return {
    ...board,
    on_the_clock: slot,
    on_the_clock_team_id: ownerOfSlot(board, slot),
  }
}

// Applicant id -> the team holding them. Patched by diffing the picks rather
// than rebuilt from the board, because this map spans the other role's board
// too and rebuilding would drop it.
export function patchDrafted(
  drafted: Record<string, string>,
  before: DraftBoard,
  after: DraftBoard
): Record<string, string> {
  const next = { ...drafted }
  const held = new Set(after.picks.map((p) => p.application_id))
  for (const pick of before.picks) {
    if (!held.has(pick.application_id)) delete next[pick.application_id]
  }
  const wasHeld = new Set(before.picks.map((p) => p.application_id))
  for (const pick of after.picks) {
    if (wasHeld.has(pick.application_id)) continue
    const team = after.teams.find((t) => t.id === pick.draft_team_id)
    if (team) next[pick.application_id] = team.name
  }
  return next
}

// Every write lands on the same board, so they all invalidate it the same way.
// The pick mutations additionally pass a `predict`, which rewrites the cached
// board so the pick is on screen before the round trip finishes — this board
// is projected in a room, and the wait was visible from the back of it.
function useDraftMutation<TVars extends { cycleId: string; role: Role }>(
  mutationFn: (vars: TVars) => Promise<unknown>,
  predict?: (board: DraftBoard, vars: TVars) => DraftBoard
) {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn,
    onMutate: predict
      ? async (vars: TVars) => {
          const boardKey = queryKeys.drafts.board(vars.cycleId, vars.role)
          const draftedKey = queryKeys.drafts.drafted(vars.cycleId)
          // Both poll every 8s; an in-flight refetch would land on top of the
          // prediction and undo it.
          await queryClient.cancelQueries({ queryKey: boardKey })
          await queryClient.cancelQueries({ queryKey: draftedKey })

          const board = queryClient.getQueryData<DraftBoard>(boardKey)
          const drafted =
            queryClient.getQueryData<Record<string, string>>(draftedKey)
          if (board) {
            const next = withOnTheClock(predict(board, vars))
            queryClient.setQueryData(boardKey, next)
            if (drafted) {
              queryClient.setQueryData(
                draftedKey,
                patchDrafted(drafted, board, next)
              )
            }
          }
          return { boardKey, draftedKey, board, drafted }
        }
      : undefined,
    onError: (_error, _vars, context) => {
      if (!context?.board) return
      queryClient.setQueryData(context.boardKey, context.board)
      if (context.drafted) {
        queryClient.setQueryData(context.draftedKey, context.drafted)
      }
    },
    // Settled rather than success: a rejected pick (someone filled the slot
    // first) has to resync as well, not just roll back to a board that is by
    // then out of date.
    onSettled: (_data, _error, vars) => {
      queryClient.invalidateQueries({
        queryKey: queryKeys.drafts.board(vars.cycleId, vars.role),
      })
      queryClient.invalidateQueries({ queryKey: queryKeys.drafts.all })
      // A pick moves the applicant to accepted (and undoing it moves them
      // back), which every application list on screen is showing.
      queryClient.invalidateQueries({ queryKey: queryKeys.applications.all })
    },
  })
}

export function useOpenDraft() {
  return useDraftMutation(
    (vars: {
      cycleId: string
      role: Role
      rounds?: number
      opts?: RequestOptions
    }) =>
      openDraft(
        vars.cycleId,
        { role: vars.role, ...(vars.rounds && { rounds: vars.rounds }) },
        vars.opts
      )
  )
}

export function useUpdateDraft() {
  return useDraftMutation(
    (vars: {
      cycleId: string
      role: Role
      draftId: string
      status?: DraftStatus
      rounds?: number
      opts?: RequestOptions
    }) =>
      updateDraft(
        vars.draftId,
        {
          ...(vars.status && { status: vars.status }),
          ...(vars.rounds && { rounds: vars.rounds }),
        },
        vars.opts
      )
  )
}

export function useSetDraftTeams() {
  return useDraftMutation(
    (vars: {
      cycleId: string
      role: Role
      draftId: string
      preferenceListIds: string[]
      opts?: RequestOptions
    }) =>
      setDraftTeams(
        vars.draftId,
        { preference_list_ids: vars.preferenceListIds },
        vars.opts
      )
  )
}

// Name and email are only for the predicted row — the caller has them from
// the pool it picked out of, and without them the board shows a bare id until
// the refetch lands.
type PickApplicant = { fullName: string; email: string }

// picked_by, previous_stage and picked_at are the server's to assign and
// nothing renders them; they're here to fill the shape until the refetch
// replaces this row wholesale.
function predictedPick(
  board: DraftBoard,
  slot: number,
  teamId: string,
  applicationId: string,
  applicant?: PickApplicant
): DraftPickDetail {
  return {
    id: `predicted-${slot}`,
    draft_id: board.id,
    draft_team_id: teamId,
    pick_number: slot,
    application_id: applicationId,
    full_name: applicant?.fullName ?? '',
    email: applicant?.email ?? '',
    picked_at: new Date().toISOString(),
    picked_by: '',
    previous_stage: '',
  }
}

export function useMakeDraftPick() {
  return useDraftMutation(
    (vars: {
      cycleId: string
      role: Role
      draftId: string
      applicationId: string
      pickNumber?: number
      applicant?: PickApplicant
      opts?: RequestOptions
    }) =>
      makeDraftPick(
        vars.draftId,
        {
          application_id: vars.applicationId,
          ...(vars.pickNumber && { pick_number: vars.pickNumber }),
        },
        vars.opts
      ),
    (board, vars) => {
      const slot = vars.pickNumber ?? board.on_the_clock
      // Nothing on the clock, or a slot that's already taken: leave the board
      // alone and let the server's answer stand rather than guess at one this
      // pick can't land in.
      if (!slot || board.picks.some((p) => p.pick_number === slot)) return board
      const teamId = ownerOfSlot(board, slot)
      if (!teamId) return board
      return {
        ...board,
        picks: [
          ...board.picks,
          predictedPick(
            board,
            slot,
            teamId,
            vars.applicationId,
            vars.applicant
          ),
        ],
      }
    }
  )
}

export function useRemoveDraftPick() {
  return useDraftMutation(
    (vars: {
      cycleId: string
      role: Role
      draftId: string
      pickNumber: number
      opts?: RequestOptions
    }) => removeDraftPick(vars.draftId, vars.pickNumber, vars.opts),
    // Emptying the slot puts it back on the clock, since it becomes the lowest
    // unfilled one.
    (board, vars) => ({
      ...board,
      picks: board.picks.filter((p) => p.pick_number !== vars.pickNumber),
    })
  )
}

// Clears every pick and returns the board to setup, putting each applicant
// back in the stage they were in before their pick. The order survives — a
// rehearsal can be wiped without rebuilding it.
export function useResetDraft() {
  return useDraftMutation(
    (vars: {
      cycleId: string
      role: Role
      draftId: string
      opts?: RequestOptions
    }) => resetDraft(vars.draftId, vars.opts)
  )
}

// Removes the board entirely, undoing its picks' stage changes first.
export function useDeleteDraft() {
  return useDraftMutation(
    (vars: {
      cycleId: string
      role: Role
      draftId: string
      opts?: RequestOptions
    }) => deleteDraft(vars.draftId, vars.opts)
  )
}

// Swaps the applicant in an already-filled slot. In place, so the slot never
// reopens and the team on the clock stays put.
export function useReplaceDraftPick() {
  return useDraftMutation(
    (vars: {
      cycleId: string
      role: Role
      draftId: string
      pickNumber: number
      applicationId: string
      applicant?: PickApplicant
      opts?: RequestOptions
    }) =>
      replaceDraftPick(
        vars.draftId,
        vars.pickNumber,
        { application_id: vars.applicationId },
        vars.opts
      ),
    // The slot keeps its number, so the board's shape and whose turn it is
    // both stay put — only the person in it changes.
    (board, vars) => ({
      ...board,
      picks: board.picks.map((p) =>
        p.pick_number === vars.pickNumber
          ? {
              ...p,
              application_id: vars.applicationId,
              full_name: vars.applicant?.fullName ?? '',
              email: vars.applicant?.email ?? '',
            }
          : p
      ),
    })
  )
}

File: backend/internal/store/drafts.go
package store

import (
	"context"
	"errors"

	"github.com/jackc/pgx/v5"

	"github.com/GenerateNU/apportal/backend/internal/models"
)

const draftColumns = `id, cycle_id, application_role, status, rounds, created_by, created_at, updated_at`

// SnakePosition maps a 1-based pick number to a 0-based seat in the order.
// Odd rounds run down the order, even rounds back up it — so the team in the
// last seat picks the end of one round and the start of the next.
func SnakePosition(pickNumber, teamCount int) int {
	if teamCount <= 0 || pickNumber <= 0 {
		return 0
	}
	index := (pickNumber - 1) % teamCount
	if round := (pickNumber-1)/teamCount + 1; round%2 == 0 {
		return teamCount - 1 - index
	}
	return index
}

func (s *Store) GetDraft(ctx context.Context, cycleID string, role models.Role) (models.Draft, error) {
	const q = `SELECT ` + draftColumns + ` FROM drafts WHERE cycle_id = $1 AND application_role = $2`
	rows, err := s.db.Query(ctx, q, cycleID, string(role))
	if err != nil {
		return models.Draft{}, err
	}
	draft, err := pgx.CollectExactlyOneRow(rows, pgx.RowToStructByPos[models.Draft])
	if errors.Is(err, pgx.ErrNoRows) {
		return models.Draft{}, ErrNotFound
	}
	return draft, err
}

func (s *Store) GetDraftByID(ctx context.Context, id string) (models.Draft, error) {
	const q = `SELECT ` + draftColumns + ` FROM drafts WHERE id = $1`
	rows, err := s.db.Query(ctx, q, id)
	if err != nil {
		return models.Draft{}, err
	}
	draft, err := pgx.CollectExactlyOneRow(rows, pgx.RowToStructByPos[models.Draft])
	if errors.Is(err, pgx.ErrNoRows) {
		return models.Draft{}, ErrNotFound
	}
	return draft, err
}

// CreateDraft opens the board for a (cycle, role). Idempotent by the table's
// unique constraint — a second call returns the existing one rather than
// erroring, since two chiefs opening the page at once shouldn't race.
func (s *Store) CreateDraft(ctx context.Context, cycleID string, role models.Role, rounds int, createdBy string) (models.Draft, error) {
	const q = `
		INSERT INTO drafts (cycle_id, application_role, rounds, created_by)
		VALUES ($1, $2, $3, $4)
		ON CONFLICT (cycle_id, application_role) DO NOTHING
		RETURNING ` + draftColumns
	rows, err := s.db.Query(ctx, q, cycleID, string(role), rounds, createdBy)
	if err != nil {
		return models.Draft{}, err
	}
	draft, err := pgx.CollectExactlyOneRow(rows, pgx.RowToStructByPos[models.Draft])
	if errors.Is(err, pgx.ErrNoRows) {
		return s.GetDraft(ctx, cycleID, role)
	}
	return draft, err
}

type DraftUpdate struct {
	Status *models.DraftStatus
	Rounds *int
}

func (s *Store) UpdateDraft(ctx context.Context, id string, in DraftUpdate) (models.Draft, error) {
	const q = `
		UPDATE drafts SET
			status = COALESCE($2, status),
			rounds = COALESCE($3, rounds)
		WHERE id = $1
		RETURNING ` + draftColumns
	var status *string
	if in.Status != nil {
		v := string(*in.Status)
		status = &v
	}
	rows, err := s.db.Query(ctx, q, id, status, in.Rounds)
	if err != nil {
		return models.Draft{}, err
	}
	draft, err := pgx.CollectExactlyOneRow(rows, pgx.RowToStructByPos[models.Draft])
	if errors.Is(err, pgx.ErrNoRows) {
		return models.Draft{}, ErrNotFound
	}
	return draft, err
}

// SetDraftTeams replaces the order wholesale: the seats are the given
// preference lists, in the given order. Teams that drop out take their picks
// with them (ON DELETE CASCADE), which is why the handler only allows this
// while the draft is still in setup.
func (s *Store) SetDraftTeams(ctx context.Context, draftID string, preferenceListIDs []string) error {
	tx, err := s.db.Begin(ctx)
	if err != nil {
		return err
	}
	defer func() { _ = tx.Rollback(ctx) }()

	const clear = `DELETE FROM draft_teams WHERE draft_id = $1 AND NOT (preference_list_id = ANY($2::uuid[]))`
	if _, err := tx.Exec(ctx, clear, draftID, preferenceListIDs); err != nil {
		return err
	}
	const upsert = `
		INSERT INTO draft_teams (draft_id, preference_list_id, position)
		SELECT $1, o.preference_list_id, o.ord - 1
		FROM unnest($2::uuid[]) WITH ORDINALITY AS o(preference_list_id, ord)
		ON CONFLICT (draft_id, preference_list_id) DO UPDATE SET position = EXCLUDED.position`
	if _, err := tx.Exec(ctx, upsert, draftID, preferenceListIDs); err != nil {
		return err
	}
	return tx.Commit(ctx)
}

func (s *Store) ListDraftTeams(ctx context.Context, draftID string) ([]models.DraftTeamDetail, error) {
	const q = `
		SELECT t.id, t.draft_id, t.preference_list_id, t.position,
		       pl.name,
		       COALESCE(
		         (SELECT array_agg(u.full_name ORDER BY u.full_name)
		          FROM preference_list_members m
		          JOIN users u ON u.nuid = m.lead_nuid
		          WHERE m.preference_list_id = pl.id),
		         '{}'
		       )
		FROM draft_teams t
		JOIN preference_lists pl ON pl.id = t.preference_list_id
		WHERE t.draft_id = $1
		ORDER BY t.position`
	rows, err := s.db.Query(ctx, q, draftID)
	if err != nil {
		return nil, err
	}
	return pgx.CollectRows(rows, pgx.RowToStructByPos[models.DraftTeamDetail])
}

func (s *Store) ListDraftPicks(ctx context.Context, draftID string) ([]models.DraftPickDetail, error) {
	const q = `
		SELECT p.id, p.draft_id, p.pick_number, p.draft_team_id, p.application_id,
		       p.previous_stage, p.picked_by, p.picked_at,
		       COALESCE(u.full_name, ''), COALESCE(u.email, '')
		FROM draft_picks p
		JOIN applications a ON a.id = p.application_id
		LEFT JOIN users u ON u.nuid = a.user_nuid
		WHERE p.draft_id = $1
		ORDER BY p.pick_number`
	rows, err := s.db.Query(ctx, q, draftID)
	if err != nil {
		return nil, err
	}
	return pgx.CollectRows(rows, pgx.RowToStructByPos[models.DraftPickDetail])
}

// DraftBoard assembles the order, the picks, and which slot is on the clock.
func (s *Store) DraftBoard(ctx context.Context, draft models.Draft) (models.DraftBoard, error) {
	board := models.DraftBoard{Draft: draft}
	teams, err := s.ListDraftTeams(ctx, draft.ID)
	if err != nil {
		return board, err
	}
	picks, err := s.ListDraftPicks(ctx, draft.ID)
	if err != nil {
		return board, err
	}
	board.Teams = teams
	board.Picks = picks
	if draft.Status != models.DraftActive || len(teams) == 0 {
		return board, nil
	}
	slot := nextOpenSlot(pickNumbers(picks), draft.Rounds*len(teams))
	if slot == 0 {
		return board, nil
	}
	board.OnTheClock = slot
	board.OnTheClockTeamID = teams[SnakePosition(slot, len(teams))].ID
	return board, nil
}

// nextOpenSlot is the lowest unfilled pick number, so undoing a pick mid-board
// puts that slot back on the clock instead of the end of the board. 0 when
// every slot is filled.
func nextOpenSlot(taken []int, total int) int {
	filled := make(map[int]bool, len(taken))
	for _, n := range taken {
		filled[n] = true
	}
	for n := 1; n <= total; n++ {
		if !filled[n] {
			return n
		}
	}
	return 0
}

func pickNumbers(picks []models.DraftPickDetail) []int {
	out := make([]int, len(picks))
	for i, p := range picks {
		out[i] = p.PickNumber
	}
	return out
}

// ErrDraftSlotTaken is returned when the requested slot already holds a pick —
// two operators submitting at once, or a stale board.
var ErrDraftSlotTaken = errors.New("draft slot already filled")

// ErrNoChange means the write asked for the state the row is already in.
var ErrNoChange = errors.New("nothing to change")

// MakeDraftPick fills the given slot and moves the applicant to accepted,
// remembering the stage they came from so an undo can put it back. Slot 0
// means "whichever is on the clock".
func (s *Store) MakeDraftPick(ctx context.Context, draft models.Draft, slot int, applicationID, pickedBy string) (models.DraftPick, error) {
	var pick models.DraftPick

	tx, err := s.db.Begin(ctx)
	if err != nil {
		return pick, err
	}
	defer func() { _ = tx.Rollback(ctx) }()

	// Serialize concurrent picks on this draft: without the lock two operators
	// can both read the same open slot and race to fill it.
	if _, err := tx.Exec(ctx, `SELECT pg_advisory_xact_lock(hashtext($1))`, draft.ID); err != nil {
		return pick, err
	}

	teams, err := s.ListDraftTeams(ctx, draft.ID)
	if err != nil {
		return pick, err
	}
	if len(teams) == 0 {
		return pick, ErrNotFound
	}
	picks, err := s.ListDraftPicks(ctx, draft.ID)
	if err != nil {
		return pick, err
	}
	total := draft.Rounds * len(teams)
	if slot == 0 {
		slot = nextOpenSlot(pickNumbers(picks), total)
	}
	if slot < 1 || slot > total {
		return pick, ErrNotFound
	}
	for _, p := range picks {
		if p.PickNumber == slot {
			return pick, ErrDraftSlotTaken
		}
	}

	// The applicant has to be in this board's cycle and role, and their stage
	// is captured here so the undo below has somewhere to put them back.
	var previousStage string
	const stageQ = `SELECT stage FROM applications WHERE id = $1 AND cycle_id = $2 AND application_role = $3`
	if err := tx.QueryRow(ctx, stageQ, applicationID, draft.CycleID, string(draft.ApplicationRole)).Scan(&previousStage); err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return pick, ErrNotFound
		}
		return pick, err
	}

	const insert = `
		INSERT INTO draft_picks (draft_id, pick_number, draft_team_id, application_id, previous_stage, picked_by)
		VALUES ($1, $2, $3, $4, $5, $6)
		RETURNING id, draft_id, pick_number, draft_team_id, application_id, previous_stage, picked_by, picked_at`
	teamID := teams[SnakePosition(slot, len(teams))].ID
	rows, err := tx.Query(ctx, insert, draft.ID, slot, teamID, applicationID, previousStage, pickedBy)
	if err != nil {
		return pick, err
	}
	pick, err = pgx.CollectExactlyOneRow(rows, pgx.RowToStructByPos[models.DraftPick])
	if err != nil {
		return pick, err
	}

	const accept = `UPDATE applications SET stage = 'accepted' WHERE id = $1`
	if _, err := tx.Exec(ctx, accept, applicationID); err != nil {
		return pick, err
	}
	return pick, tx.Commit(ctx)
}

// ErrDraftAlreadyPicked is returned when the applicant is already claimed in
// another slot on the same board.
var ErrDraftAlreadyPicked = errors.New("applicant already drafted on this board")

// ReplaceDraftPick swaps the applicant in a slot that's already filled — the
// team changed its mind. Done in place rather than as a remove plus a re-pick,
// so the slot never opens: an empty slot mid-board is the one on the clock,
// which would drag the draft backwards while everyone waits.
//
// The outgoing applicant goes back to the stage they were in, and the incoming
// one's stage is captured in its place, so undo still works afterwards.
func (s *Store) ReplaceDraftPick(ctx context.Context, draft models.Draft, pickNumber int, applicationID, pickedBy string) (models.DraftPick, error) {
	var pick models.DraftPick

	tx, err := s.db.Begin(ctx)
	if err != nil {
		return pick, err
	}
	defer func() { _ = tx.Rollback(ctx) }()

	if _, err := tx.Exec(ctx, `SELECT pg_advisory_xact_lock(hashtext($1))`, draft.ID); err != nil {
		return pick, err
	}

	var oldApplicationID, oldPreviousStage string
	const current = `
		SELECT application_id, previous_stage FROM draft_picks
		WHERE draft_id = $1 AND pick_number = $2 FOR UPDATE`
	if err := tx.QueryRow(ctx, current, draft.ID, pickNumber).Scan(&oldApplicationID, &oldPreviousStage); err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return pick, ErrNotFound
		}
		return pick, err
	}
	if oldApplicationID == applicationID {
		return pick, ErrNoChange
	}

	// Claimed in another slot on this board — the unique constraint would
	// catch it, but a named conflict tells the operator what actually happened.
	var takenElsewhere bool
	const taken = `
		SELECT EXISTS (SELECT 1 FROM draft_picks
		               WHERE draft_id = $1 AND application_id = $2 AND pick_number <> $3)`
	if err := tx.QueryRow(ctx, taken, draft.ID, applicationID, pickNumber).Scan(&takenElsewhere); err != nil {
		return pick, err
	}
	if takenElsewhere {
		return pick, ErrDraftAlreadyPicked
	}

	var newPreviousStage string
	const stageQ = `SELECT stage FROM applications WHERE id = $1 AND cycle_id = $2 AND application_role = $3`
	if err := tx.QueryRow(ctx, stageQ, applicationID, draft.CycleID, string(draft.ApplicationRole)).Scan(&newPreviousStage); err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return pick, ErrNotFound
		}
		return pick, err
	}

	// Put the outgoing applicant back before the incoming one is accepted, so
	// a swap between two rows can't leave both in the same stage.
	const restore = `UPDATE applications SET stage = $2 WHERE id = $1 AND stage = 'accepted'`
	if _, err := tx.Exec(ctx, restore, oldApplicationID, oldPreviousStage); err != nil {
		return pick, err
	}

	const update = `
		UPDATE draft_picks
		SET application_id = $3, previous_stage = $4, picked_by = $5, picked_at = NOW()
		WHERE draft_id = $1 AND pick_number = $2
		RETURNING id, draft_id, pick_number, draft_team_id, application_id, previous_stage, picked_by, picked_at`
	rows, err := tx.Query(ctx, update, draft.ID, pickNumber, applicationID, newPreviousStage, pickedBy)
	if err != nil {
		return pick, err
	}
	pick, err = pgx.CollectExactlyOneRow(rows, pgx.RowToStructByPos[models.DraftPick])
	if err != nil {
		return pick, err
	}

	const accept = `UPDATE applications SET stage = 'accepted' WHERE id = $1`
	if _, err := tx.Exec(ctx, accept, applicationID); err != nil {
		return pick, err
	}
	return pick, tx.Commit(ctx)
}

// RemoveDraftPick empties a slot and puts the applicant back in the stage they
// were in before it — but only if they're still in accepted, so a stage a
// chief moved deliberately afterwards isn't clobbered.
func (s *Store) RemoveDraftPick(ctx context.Context, draftID string, pickNumber int) error {
	tx, err := s.db.Begin(ctx)
	if err != nil {
		return err
	}
	defer func() { _ = tx.Rollback(ctx) }()

	var applicationID, previousStage string
	const del = `
		DELETE FROM draft_picks WHERE draft_id = $1 AND pick_number = $2
		RETURNING application_id, previous_stage`
	if err := tx.QueryRow(ctx, del, draftID, pickNumber).Scan(&applicationID, &previousStage); err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return ErrNotFound
		}
		return err
	}

	const restore = `UPDATE applications SET stage = $2 WHERE id = $1 AND stage = 'accepted'`
	if _, err := tx.Exec(ctx, restore, applicationID, previousStage); err != nil {
		return err
	}
	return tx.Commit(ctx)
}

// restoreDraftedStages puts every applicant a draft claimed back in the stage
// they were in before the pick. Scoped to rows still sitting in accepted, so a
// stage someone moved deliberately after the pick isn't clobbered — the same
// rule RemoveDraftPick follows for one pick.
func restoreDraftedStages(ctx context.Context, tx pgx.Tx, draftID string) error {
	const q = `
		UPDATE applications a SET stage = p.previous_stage
		FROM draft_picks p
		WHERE p.draft_id = $1 AND a.id = p.application_id AND a.stage = 'accepted'`
	_, err := tx.Exec(ctx, q, draftID)
	return err
}

// ResetDraft clears every pick and returns the board to setup, undoing the
// stage changes on the way out. The team order survives, so a rehearsal can be
// wiped without rebuilding it.
func (s *Store) ResetDraft(ctx context.Context, draftID string) error {
	tx, err := s.db.Begin(ctx)
	if err != nil {
		return err
	}
	defer func() { _ = tx.Rollback(ctx) }()

	if err := restoreDraftedStages(ctx, tx, draftID); err != nil {
		return err
	}
	if _, err := tx.Exec(ctx, `DELETE FROM draft_picks WHERE draft_id = $1`, draftID); err != nil {
		return err
	}
	tag, err := tx.Exec(ctx, `UPDATE drafts SET status = 'setup' WHERE id = $1`, draftID)
	if err != nil {
		return err
	}
	if tag.RowsAffected() == 0 {
		return ErrNotFound
	}
	return tx.Commit(ctx)
}

// DeleteDraft removes the board outright. Picks cascade, so their stage
// changes are undone first — otherwise the applicants would be left in
// accepted with nothing on record explaining why.
func (s *Store) DeleteDraft(ctx context.Context, draftID string) error {
	tx, err := s.db.Begin(ctx)
	if err != nil {
		return err
	}
	defer func() { _ = tx.Rollback(ctx) }()

	if err := restoreDraftedStages(ctx, tx, draftID); err != nil {
		return err
	}
	tag, err := tx.Exec(ctx, `DELETE FROM drafts WHERE id = $1`, draftID)
	if err != nil {
		return err
	}
	if tag.RowsAffected() == 0 {
		return ErrNotFound
	}
	return tx.Commit(ctx)
}

// DraftedApplicationIDs is every application already claimed on any board in
// the cycle, for greying them out on the preference lists people draft from.
func (s *Store) DraftedApplicationIDs(ctx context.Context, cycleID string) (map[string]string, error) {
	const q = `
		SELECT p.application_id, pl.name
		FROM draft_picks p
		JOIN drafts d       ON d.id = p.draft_id
		JOIN draft_teams t  ON t.id = p.draft_team_id
		JOIN preference_lists pl ON pl.id = t.preference_list_id
		WHERE d.cycle_id = $1`
	rows, err := s.db.Query(ctx, q, cycleID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := map[string]string{}
	for rows.Next() {
		var applicationID, teamName string
		if err := rows.Scan(&applicationID, &teamName); err != nil {
			return nil, err
		}
		out[applicationID] = teamName
	}
	return out, rows.Err()
}
